process/track/shardBlockTrack.go

```diff
@@ -0,0 +1,135 @@
+	}
+
+	crossNotarizer, err := NewBlockNotarizer(arguments.Hasher, arguments.Marshalizer)
+	if err != nil {
+		return nil, err
+	}
+
+	selfNotarizer, err := NewBlockNotarizer(arguments.Hasher, arguments.Marshalizer)
+	if err != nil {
+		return nil, err
+	}
+
+	crossNotarizedHeadersNotifier, err := NewBlockNotifier()
+	if err != nil {
+		return nil, err
+	}
+
+	selfNotarizedHeadersNotifier, err := NewBlockNotifier()
+	if err != nil {
+		return nil, err
+	}
+
+	bbt := &baseBlockTrack{
+		hasher:                        arguments.Hasher,
+		headerValidator:               arguments.HeaderValidator,
+		marshalizer:                   arguments.Marshalizer,
+		rounder:                       arguments.Rounder,
+		shardCoordinator:              arguments.ShardCoordinator,
+		shardHeadersPool:              arguments.PoolsHolder.Headers(),
+		metaBlocksPool:                arguments.PoolsHolder.MetaBlocks(),
+		headersNoncesPool:             arguments.PoolsHolder.HeadersNonces(),
+		store:                         arguments.Store,
+		crossNotarizer:                crossNotarizer,
+		selfNotarizer:                 selfNotarizer,
+		crossNotarizedHeadersNotifier: crossNotarizedHeadersNotifier,
+		selfNotarizedHeadersNotifier:  selfNotarizedHeadersNotifier,
+	}
+
+	err = bbt.initNotarizedHeaders(arguments.StartHeaders)
+	if err != nil {
+		return nil, err
+	}
+
+	sbt := shardBlockTrack{
+		baseBlockTrack: bbt,
+	}
+
+	blockProcessor, err := NewBlockProcessor(
+		arguments.HeaderValidator,
+		arguments.ShardCoordinator,
+		&sbt,
+		crossNotarizer,
+		crossNotarizedHeadersNotifier,
+		selfNotarizedHeadersNotifier,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	sbt.blockProcessor = blockProcessor
+
+	sbt.headers = make(map[uint32]map[uint64][]*headerInfo)
+	sbt.shardHeadersPool.RegisterHandler(sbt.receivedShardHeader)
+	sbt.metaBlocksPool.RegisterHandler(sbt.receivedMetaBlock)
+
+	return &sbt, nil
+}
+
+func (sbt *shardBlockTrack) getSelfHeaders(headerHandler data.HeaderHandler) []*headerInfo {
+	selfHeadersInfo := make([]*headerInfo, 0)
+
+	metaBlock, ok := headerHandler.(*block.MetaBlock)
+	if !ok {
+		log.Debug("getSelfHeaders", process.ErrWrongTypeAssertion)
+		return selfHeadersInfo
+	}
+
+	for _, shardInfo := range metaBlock.ShardInfo {
+		if shardInfo.ShardID != sbt.shardCoordinator.SelfId() {
+			continue
+		}
+
+		header, err := process.GetShardHeader(shardInfo.HeaderHash, sbt.shardHeadersPool, sbt.marshalizer, sbt.store)
+		if err != nil {
+			log.Debug("GetShardHeader", err.Error())
+			continue
+		}
+
+		selfHeadersInfo = append(selfHeadersInfo, &headerInfo{hash: shardInfo.HeaderHash, header: header})
+	}
+
+	return selfHeadersInfo
+}
+
+func (sbt *shardBlockTrack) computeLongestSelfChain() (data.HeaderHandler, []byte, []data.HeaderHandler, [][]byte) {
+	lastSelfNotarizedHeader, lastSelfNotarizedHeaderHash, err := sbt.selfNotarizer.getLastNotarizedHeader(sharding.MetachainShardId)
+	if err != nil {
+		return nil, nil, nil, nil
```

Comment:
treat the error.

Reply:
it is critical if that returns error. should not ever happen. In that case there should be a node restart.

Reply:
I have added a warn log. Actually getLastNotarizedHeader could not return err as notarizedHeaders map is initialized with startHeaders(genesis header for each shard) on New.. method.This method returns the new finalized headers constructed above the last finalized one in self chain. If the chain didn't advance and the node is sync or the last notarized header could not be determined, this method has nothing to return. Its return is used only to notify the registered handlers, if they will exist, that a new finalized header(s) in self chain has/have appeared.